codice/ddf
# [2.13.x] DDF-4133 Solr auth changes

##### ABBREVIATED REVIEW BETWEEN 2.13.X AND MASTER IS IN EFFECT
Porting changes from master branch to release (2.13.x) branch.
Link to 3 master branch PRs:
- https://github.com/codice/ddf/pull/4083
- https://github.com/codice/ddf/pull/4142
- https://github.com/codice/ddf/pull/4160

The master branch PRs are based on changes from PR that breaks up the solr-dependencies JAR: https://github.com/codice/ddf/pull/4134. That PR affects many, many additional files and was not backported. Therefore, there are some minor differences between the master and release branch implementations.

#### What does this PR do?
- Changes the DDF Solr connection from two-way TLS to one-way TLS
- Adds basic authentication for the DDF to prove its identity to Solr
- Adds a module to generate a basic auth password, change the password in Solr, encrypt the password, and save the encrypted password to a file.
- Introduces new system properties to control behavior (see testing section)

#### Who is reviewing it? 
@austinsteffes 
@tyler30clemens 
#### Select relevant component teams: 
@codice/security 
@codice/solr 

#### Ask 2 committers to review/merge the PR and tag them here.
@brjeter
@vinamartin

#### How should this be tested?
1. Review documentation changes in section 7.5.1.2.2. Solr Server Password Management of the documentation. Perform the configuration task described in the section. Confirm that the system behaves as documented.
1. Change the values or combination of values of the new system properties and determine if they work as described in the documentation section Managed Solr Server Properties
solr.useBasicAuth
solr.username
solr.password
solr.attemptAutoPasswordChange

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
